### PR TITLE
Job e2e

### PR DESCRIPTION
**Pull request by erictune**
This supercedes #14354

I rebased that PR, and made some test improvements.

I am not squashing, in order to respect original authorship.   Ignore CLA complaints.  We both signed it.

### CODE REVIEW

**Comment by erictune** [comment]
@mikedanese suggested reviewer

**Comment by soltysh** [comment]
@erictune I really like the changes you've done to the original tests. One tiny nit but LGTM.

**Comment by erictune** [comment]
@soltysh thanks, fixed.

**Comment by erictune** [comment]
Okay, this is passing on my GCE cluster.  :pray: that is passes on jenkins

File: test/e2e/job.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	lotsOfFailures := 20 // considerably more than completions
+
+	// Simplest case: all pods succeed promtly
+	It("should run a job to completion when tasks succeed", func() {
+		By("Creating a job")
+		job := newTestJob("succeed", "all-succeed", api.RestartPolicyNever, parallelism, completions)
+		job, err := createJob(f.Client, f.Namespace.Name, job)
+		Expect(err).NotTo(HaveOccurred())
+		// Cleanup jobs when we are done.
+		defer func() {
```
Comment (mikedanese) [inline test/e2e/job.go]: This is not strictly necessary. The Framework deletes the namespace and the namespace controller should then cleanup these jobs.

File: test/e2e/job.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should stop a job", func() {
+		By("Creating a job")
+		job := newTestJob("notTerminate", "foo", api.RestartPolicyNever, parallelism, completions)
+		job, err := createJob(f.Client, f.Namespace.Name, job)
+		Expect(err).NotTo(HaveOccurred())
+		// Cleanup jobs when we are done.
+		defer func() {
```
Comment (mikedanese) [inline test/e2e/job.go]: This delete will always fail with not found since we reap the job below

File: test/e2e/job.go (hunk 1 above)
Comment (mikedanese) [inline test/e2e/job.go]: #14666

File: pkg/client/unversioned/jobs.go [hunk 3]
```diff
@@ -85,7 +89,7 @@ func (c *jobs) Delete(name string, options *api.DeleteOptions) (err error) {
 		return c.r.Delete().Namespace(c.ns).Resource("jobs").Name(name).Do().Error()
 	}
 
-	body, err := api.Scheme.EncodeToVersion(options, c.r.APIVersion())
+	body, err := api.Scheme.EncodeToVersion(options, latest.GroupOrDie("").GroupVersion)
```
Comment (mikedanese) [inline pkg/client/unversioned/jobs.go]: Why is this change necessary? Scanning other client implementations I see `c.client.APIVersion()` or `c.r.APIVersion()` but never `latest.GroupOrDie`

Comment (erictune) [inline pkg/client/unversioned/jobs.go]: $ ./kubectl delete jobs/pi
W0928 12:12:13.210531   71463 helper.go:232] Server does not support API version 'experimental/v1alpha1'. Falling back to 'v1'.

Comment (erictune) [inline pkg/client/unversioned/jobs.go]: I think it is because jobs is the only experimental resource that uses DeleteOpts?

Comment (mikedanese) [inline pkg/client/unversioned/jobs.go]: Horizontal Pod Autoscaler is also experimental.

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/horizontalpodautoscaler.go#L76

@caesarxuchao can you advise?

Comment (caesarxuchao) [inline pkg/client/unversioned/jobs.go]: I think the horizontalpodautoscaler code will also fail if that line is executed. I think Eric's fix is a temporary one. In the future, apart from the ExperimentalClient, struct like ehorizontalpodautoscaler shall contain a RESTClient as well.

Comment (erictune) [inline pkg/client/unversioned/jobs.go]: The error is `no kind "DeleteOptions" is registered for version "experimental/v1alpha1"
`, not the one I said above.

File: test/e2e/job.go (hunk 2 above)
Comment (erictune) [inline test/e2e/job.go]: deleted

File: test/e2e/job.go (hunk 1 above)
Comment (erictune) [inline test/e2e/job.go]: ok, deleted all those defers.

**Comment by erictune** [comment]
passes locally.

**Comment by mikedanese** [comment]
@k8s-bot ok to test

File: test/e2e/job.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// How long to wait for a job to finish.
+	jobTimeout = 5 * time.Minute
```
Comment (mikedanese) [inline test/e2e/job.go]: This timeout seems a bit excessive.

**Comment by mikedanese** [comment]
LGTM

**Comment by bprashanth** [comment]
Out of curiosity what suite/test-list are you planning to run this in?

**Comment by erictune** [comment]
Not planning any special handling with regards to suite.

**Comment by erictune** [comment]
@k8s-oncall can this merge?  It tests a feature that is in v1.1 and needs to be merged so it can be cherry-picked to v1.1, so that v1.1 can be tested.
